### PR TITLE
XMR: HP15 support, implement BulletproofPlus, refactor and cleanup

### PR DESCRIPTION
**Pull request by ph4r05**
Changes:

- Add HP15 support
- implement BulletProof plus verifier and prover. BP+ will be mandatory in upcoming hard-fork.
- use bulletproof exception to signalize proof generation failed and should be tried again. More robust, fixes bug that was not triggered yet (return tuple did not work properly in all situations)
- precomputed 2**i vector is removed as it can be easily computed on the fly
- BP code cleanup, minor optimizations, comments

Based on https://github.com/trezor/trezor-firmware/pull/2219

A new [hardfork is being prepared](https://www.getmonero.org/2022/04/20/network-upgrade-july-2022.html), part of which is a new [Bulletproof+](https://eprint.iacr.org/2020/735.pdf) protocol. 

Other related info:
- https://github.com/monero-project/meta/issues/630
- https://github.com/monero-project/meta/issues/690

### Technical note
I had to add new BP+ `GI`, `HI` constants. This increases firmware memory footprint by 16kB, TWO vector was removed, so total 14 kB increase. Until the new hard-fork is activated, we have to keep both old and new constants to make computation possible. If there is a problem with memory, we can shorten older BP constants by half (which causes longer generation times as the rest has to be costly computed on-the-fly). In that case the total extra size would be 6144 B.

### CODE REVIEW

**Comment by ph4r05** [comment]
From some reason I cannot get rid of the styling errors. Any ideas @matejcik @grdddj ? :)

**Comment by ph4r05** [comment]
When I tried to install firmware with this new BP+ version the image was too big to fit to the flash memory. For now I removed upper 4096B from old `GI`, `HI` vectors. Everything works the same, but verification time for larger transactions will take longer

**Comment by ph4r05** [comment]
Interestingly, after device starts, I get HP internal error with "please, unplug the device" error message. Ideas how to diagnose it?

**Comment by ph4r05** [comment]
Another thing - I need to update protobuf defs in order to support new HF. But it needs to be merged so I can open Monero Wallet PR. Any ideas on a timeframe for the PR being merged? :)

**Comment by ph4r05** [comment]
OK I solved it

**Comment by sethforprivacy** [comment]
Thank you so much for all of the hard work on this, @ph4r05! Great to see such top-notch support for Monero on Trezor.

**Comment by ph4r05** [comment]
Thanks @sethforprivacy! It was really fun implementing Bulletproof+ for Trezor :) Also thanks for pinging us ahead of time

**Comment by ph4r05** [comment]
Another thing, I hard-coded to the monero wallet that firmware version 2.5.1 supports HF15. Should a new firmware with the PR included have higher version, pls let me know so I change it in the monero PR. thanks!

**Comment by matejcik** [comment]
> Another thing, I hard-coded to the monero wallet that firmware version 2.5.1 supports HF15. Should a new firmware with the PR included have higher version, pls let me know so I change it in the monero PR. thanks!

💩  didn't notice this one.
We are releasing 2.5.1 which does not include this PR. Please bump to 2.5.2 in the monero PR.

File: core/src/apps/monero/xmr/mlsag_hasher.py [hunk 1]
```diff
@@ -74,22 +75,30 @@ def rsig_val(self, p: bytes | list[bytes] | Bulletproof, raw: bool = False) -> N
...
-        self.rsig_hasher.update(p.mu)
+        fields = (
+            (p.A, p.A1, p.B, p.r1, p.s1, p.d1)
+            if is_plus
+            else (p.A, p.S, p.T1, p.T2, p.taux, p.mu)
+        )
+        for fld in fields:
+            self.rsig_hasher.update(fld)
+
+        del (fields,)
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply core/src/apps/monero/xmr/mlsag_hasher.py:92]: As discussed, I will resolve it, ok?

**Comment by bosomt** [comment]
@ph4r05 Is there anything that can/need to be tested on our side for 2.5.2 or discovery/send/receive of TX is enough ?

https://www.exploremonero.com/transaction/d605e8a1417035aceb60514c4cea54feb3e9af8dc740e93ac220379a79f33531


`Device: model T 2.5.2 Universal (revision 0d87b55ba4fed7eecc72bf2a94ee473830b095e9)`

**Comment by matejcik** [comment]
it will be enough, but afer HF15 activates on Aug 13 https://www.getmonero.org/2022/04/20/network-upgrade-july-2022.html

**Comment by bosomt** [comment]
**QA OK**

 - moneroGUI 0.18.1.0-release (Qt 5.12.8)
 - Device: model T 2.5.2 regular (revision 0d87b55ba4fed7eecc72bf2a94ee473830b095e9)
- txid 1c45ce86d6ade9d7574a93e4514d8dccee909195c7b1ea0bc5a6ad54919e8a98

image

**Comment by bosomt** [comment]
One question, im not sure ho MoneroGUI works but i rather ask if its OK.
When i press show address on device i see completely different address :) 
But it belongs to same account /i just sent some monero to that address and i still have same amount ;) 

Screenshot 2022-08-17 at 11 04 29

image

**Comment by prusnak** [comment]
> When i press show address on device i see completely different address :)

Please open another issue for this

**Pull request merged**